code/obj/mining_cloud_storage.dm

```diff
@@ -0,0 +1,372 @@
+				user.u_equip(R)
+				R.dropped()
+		else if(R.amount>1)
+			R.set_loc(src)
+			for(R.amount,R.amount > 0, R.amount--)
+				var/obj/item/raw_material/new_mat = unpool(R.type)
+				new_mat.loc = src
+				amount_loaded++
+			if (user && R)
+				user.u_equip(R)
+				R.dropped()
+			pool(R)
+		update_ore_amount(R.material_name,amount_loaded)
+
+
+	proc/accept_loading(var/mob/user,var/allow_silicon = 0)
+		if (!user)
+			return 0
+		if (src.status & BROKEN || src.status & NOPOWER)
+			return 0
+		if (!istype(user, /mob/living/))
+			return 0
+		if (istype(user, /mob/living/silicon) && !allow_silicon)
+			return 0
+		var/mob/living/L = user
+		if (!isalive(L) || L.transforming)
+			return 0
+		return 1
+
+	proc/update_ore_amount(var/material_name,var/delta)
+		if(ores[material_name])
+			var/datum/ore_cloud_data/OCD = ores[material_name]
+			OCD.amount += delta
+			OCD.amount = max(OCD.amount,0)
+			ores[material_name] = OCD
+		else if (delta > 0)
+			ores += material_name
+			var/datum/ore_cloud_data/OCD = new /datum/ore_cloud_data
+			OCD.amount += delta
+			OCD.for_sale = 0
+			OCD.price = 0
+			ores[material_name] = OCD
+
+	proc/update_ore_for_sale(var/material_name,var/new_for_sale)
+		if(ores[material_name])
+			var/datum/ore_cloud_data/OCD = ores[material_name]
+			if(isnull(new_for_sale))
+				OCD.for_sale = !OCD.for_sale
+			else
+				OCD.for_sale = new_for_sale
+			ores[material_name] = OCD
+		return
+
+	proc/update_ore_price(var/material_name,var/new_price)
+		if(ores[material_name])
+			var/datum/ore_cloud_data/OCD = ores[material_name]
+			OCD.price = max(0,new_price)
+			ores[material_name] = OCD
+		return
+
+	attack_hand(var/mob/user as mob)
+
+		src.add_dialog(user)
+
+		var/dat = ""
+
+		if (status & BROKEN || status & NOPOWER)
+			dat = "The screen is blank."
+			user << browse(dat, "window=mining_dropbox;size=400x500")
+			onclose(user, "mining_dropbox")
+			return
+
+		dat = "<B>[src.name]</B>"
+
+		dat += "<br><HR>"
+
+		dat += "<B>Rockbox&trade; Ore Cloud Storage Service Settings:</B>"
+
+		dat += "<br><small>"
+
+		dat += "<B>Rockbox&trade; Fees:</B> $[!rockbox_globals.rockbox_premium_purchased ? ROCKBOX_STANDARD_FEE : 0] per ore [!rockbox_globals.rockbox_premium_purchased ? "(Purchase our Premium Service to remove this fee!)" : ""]<BR>"
+		dat += "<B>Client Quartermaster Transaction Fee:</B> [rockbox_globals.rockbox_client_fee_pct]%<BR>"
+		dat += "<B>Client Quartermaster Transaction Fee Per Ore Minimum:</B> $[rockbox_globals.rockbox_client_fee_min]<BR>"
+
+		dat += "</small><HR>"
+
+		if(ores.len)
+			for(var/ore in ores)
+				var/sellable = 0
+				var/price = 0
+				var/datum/ore_cloud_data/OCD = ores[ore]
+				if(OCD)
+					price = OCD.price
+					sellable = OCD.for_sale
+				dat += "<B>[ore]:</B> [OCD.amount] (<A href='?src=\ref[src];sellable=[ore]'>[sellable ? "For Sale" : "Not For Sale"]</A>) (<A href='?src=\ref[src];price=[ore]'>$[price] per ore</A>) (<A href='?src=\ref[src];eject=[ore]'>Eject</A>)<br>"
```

Comment:
```suggestion
			for(var/ore in ores)
				var/datum/ore_cloud_data/OCD = ores[ore]
				var/price = OCD.price
				var/sellable = OCD.for_sale
```
If `ore` in `ores` exists, `ores[ore]` must exist, right?

Reply:
Not necessarily; while the code I've written shouldn't allow for it I think a runtime or some future thing I've not accounted for could potentially lead to it being in there with the datum missing

Are we good with not being paranoid about that?

Reply:
I think so, yes.

Reply:
implemented in pending commit